src/Layers/xrRenderDX11/dx11SH_Texture.cpp

```diff
@@ -117,23 +117,17 @@ void CTexture::surface_set(ID3DBaseTexture* surf)
             _RELEASE(srv_all);
             CHK_DX(HW.pDevice->CreateShaderResourceView(pSurface, &ViewDesc, &srv_all));
 
+            if (desc.SampleDesc.Count <= 1)
+                ViewDesc.Texture2DArray.ArraySize = desc.ArraySize;
+            else
+                ViewDesc.Texture2DMSArray.ArraySize = desc.ArraySize;
+            
             srv_per_slice.resize(desc.ArraySize);
-            for (u32 id = 0; id < desc.ArraySize; ++id)
-            {
-                _RELEASE(srv_per_slice[id]);
+            for (auto &x : srv_per_slice)
+                _RELEASE(x);
+
+            CHK_DX(HW.pDevice->CreateShaderResourceView(pSurface, &ViewDesc, srv_per_slice.data()));
```

Comment:
I am an absolute lamer in matters of rendering and the DX11 api, but It seems like in can be simplified in this way

Reply:
No, this was made on purpose – SRV per slice. It basically makes the engine work with original game shaders, because they were written to work with one texture, not texture arrays.

So, your change break the game :)

Reply:
I played with "release master gold" build for ~15 minutes and it didn't break. 
But in mixed build type it feels like it decreases performance for ~3-5fps.

Doesn't this call just create a shader payload that should be equal in the end?

Reply:
well, ok, I see this now
![image](https://github.com/OpenXRay/xray-16/assets/68744350/2b303588-0d4b-4286-93b3-0a5320f2e64a)

Reply:
It doesn't create the same payload.
We want to create a number of SRVs, where each created SRV contains only one texture (slice) from a texture array srv_all.
With the this change, only one SRV that contains an entire texture array with all textures instead of just one texture is being created.

I though, that this just breaks shaders, but it seems that it's different.

Reply:
I thought that would just break the shaders, but it behaves differently: D3D11 runtime just always binds srv_per_slice[0] during the shaders launch, where we want to use different slices instead of first.

Reply:
Ok, thanks. By the way, do you have an idea how these calls could be handled concurrently if I did something like this:
```c++
xr_parallel_for(TaskRange<u32>(0, desc.ArraySize), [this, ViewDesc] (const auto& range) {
    D3D_SHADER_RESOURCE_VIEW_DESC ViewDescLocal = ViewDesc; 
    UINT* FirstArraySlice = &ViewDescLocal.Texture2DMSArray.FirstArraySlice;
    if (desc.SampleDesc.Count <= 1)
        FirstArraySlice = &ViewDescLocal.Texture2DArray.FirstArraySlice;
    
    for (u32 id = range.begin(); id != range.end(); ++id)
    {
        _RELEASE(srv_per_slice[id]);
        *FirstArraySlice = id;
        CHK_DX(HW.pDevice->CreateShaderResourceView(pSurface, &ViewDescLocal, &srv_per_slice[id]));
    }
});
```

Reply:
And is rewriting shaders is an option?

Reply:
> And is rewriting shaders is an option?

Yes and no.
We can have modified shaders but still must maintain compatibility with vanilla shaders.



> Ok, thanks. By the way, do you have an idea how these calls could be handled concurrently if I did something like this:
> 
> ```c++
> xr_parallel_for(TaskRange<u32>(0, desc.ArraySize), [this, ViewDesc] (const auto& range) {
>     D3D_SHADER_RESOURCE_VIEW_DESC ViewDescLocal = ViewDesc; 
>     UINT* FirstArraySlice = &ViewDescLocal.Texture2DMSArray.FirstArraySlice;
>     if (desc.SampleDesc.Count <= 1)
>         FirstArraySlice = &ViewDescLocal.Texture2DArray.FirstArraySlice;
>     
>     for (u32 id = range.begin(); id != range.end(); ++id)
>     {
>         _RELEASE(srv_per_slice[id]);
>         *FirstArraySlice = id;
>         CHK_DX(HW.pDevice->CreateShaderResourceView(pSurface, &ViewDescLocal, &srv_per_slice[id]));
>     }
> });
> ```

I'm not sure if `CreateShaderResourceView` is thread-safe, but fixing this code with the same approach is still suboptimal.
Normally, we should not call `CreateShaderResourceView` every frame, which we are exactly doing currently.
And it's better to implement that change in a separate pull request if you attempt to do it.

@vTurbine might tell more

Reply:
> And it's better to implement that change in a separate pull request if you attempt to do it.

That will be a major code rewrite, so I'm not sure how much efforts it will take.

Reply:
> Ok, thanks. By the way, do you have an idea how these calls could be handled concurrently if I did something like this:
> 
> ```c++
> xr_parallel_for(TaskRange<u32>(0, desc.ArraySize), [this, ViewDesc] (const auto& range) {
>     D3D_SHADER_RESOURCE_VIEW_DESC ViewDescLocal = ViewDesc; 
>     UINT* FirstArraySlice = &ViewDescLocal.Texture2DMSArray.FirstArraySlice;
>     if (desc.SampleDesc.Count <= 1)
>         FirstArraySlice = &ViewDescLocal.Texture2DArray.FirstArraySlice;
>     
>     for (u32 id = range.begin(); id != range.end(); ++id)
>     {
>         _RELEASE(srv_per_slice[id]);
>         *FirstArraySlice = id;
>         CHK_DX(HW.pDevice->CreateShaderResourceView(pSurface, &ViewDescLocal, &srv_per_slice[id]));
>     }
> });
> ```

Apparently, this only has a negative effect

before:
![image](https://github.com/OpenXRay/xray-16/assets/68744350/78358a91-57fb-400b-a018-816874887ae1)

after:
![image](https://github.com/OpenXRay/xray-16/assets/68744350/64179423-8ece-4ba3-be35-e569c2c01024)

It seems like `CreateShaderResourceView` is threadsafe, but not concurrent